pkg/tfshim/sdk-v2/provider_diff.go

```diff
@@ -165,7 +165,28 @@ func simpleDiffViaPlanState(
 	if state.RawConfig.IsNull() {
 		state.RawConfig = rawConfigVal
 	}
-	return res.SimpleDiff(ctx, state, planned, meta)
+	diff, err := res.SimpleDiff(ctx, state, planned, meta)
+	if err != nil {
+		return nil, err
+	}
+
+	// TF gRPC servers compensate for the fact that SimpleDiff may return
+	// terraform.NewInstanceDiff(), nil dropping any available information on RawPlan.
+	//
+	// See for example this code in PlanResourceChange:
+	//
+	// See https://github.com/hashicorp/terraform-plugin-sdk/blob/
+	//         28e631776d97f0a5a5942b3524814addbef90875/helper/schema/grpc_provider.go#L797
+	//
+	// In TF this is communicated from PlanResourceChange to ApplyResourceChange; unlike TF, in
+	// the current codebase InstanceDiff is passed directly to Apply. If RawPlan is not set on
+	// the diff it may cause nil panics in the provider.
+	if diff != nil && len(diff.Attributes) == 0 {
```

Comment:
I don't understand why it is impossible that `diff.RawPlan` will hold some other useful value when `len(diff.Attributes) == 0`. That said this mirrors upstream, so I'm inclined to trust it.

Reply:
Yup. I think it may be assuming that "when there are no changes", the plan is to keep things as -is, so PLAN = STATE. But it might be overriding a value in RawPlan that's non-nil sometimes. Hopefully that value was already equal to STATE.